Bottom: 05bb3ef91175344d488c9c73b63f40068f9250d1
Top:    262288fbc6b85d28b7a8023e87fa7bc72a99156e
Author: Peter Clifton <peter@clifton-electronics.co.uk>
Date:   2016-03-12 22:20:36 +0000

Refresh of fixup-searchobjectbylocation

---

diff --git a/src/search.c b/src/search.c
index 8d7178e..6e4f02e 100644
--- a/src/search.c
+++ b/src/search.c
@@ -1130,64 +1130,50 @@ IsPointOnArc (Coord X, Coord Y, Coord Radius, ArcType *Arc)
 }
 
 static unsigned int
-SearchLayerObjectInternal (LayerType *SearchLayer,
-                           LayerType **Result1,
+SearchLayerObjectInternal (LayerType *layer,
+                           unsigned int Type,
+                           unsigned int HigherAvail,
+                           double HigherBound,
+                           void **Result1,
                            void **Result2,
                            void **Result3)
 {
-  if (!SearchLayer->On)
+  int locked = Type & LOCKED_TYPE;
+
+  if (!layer->On)
     return NO_TYPE;
 
+  SearchLayer = layer;
+
   if ((HigherAvail & (PIN_TYPE | PAD_TYPE)) == 0 &&
       Type & POLYGONPOINT_TYPE &&
-      SearchPointByLocation (locked,
-                             Result1,
-                             (PolygonType **) Result2,
-                             (PointType **) Result3))
-    return (POLYGONPOINT_TYPE);
+      SearchPointByLocation (locked, (LayerType **)Result1, (PolygonType **) Result2, (PointType **) Result3))
+    return POLYGONPOINT_TYPE;
 
   if ((HigherAvail & (PIN_TYPE | PAD_TYPE)) == 0 &&
       Type & LINEPOINT_TYPE &&
-      SearchLinePointByLocation (locked,
-                                 Result1,
-                                 (LineType **) Result2,
-                                 (PointType **) Result3))
-    return (LINEPOINT_TYPE);
+      SearchLinePointByLocation (locked, (LayerType **)Result1, (LineType **) Result2, (PointType **) Result3))
+    return LINEPOINT_TYPE;
 
   if ((HigherAvail & (PIN_TYPE | PAD_TYPE)) == 0 && Type & LINE_TYPE
-      && SearchLineByLocation (locked,
-                               Result1,
-                               (LineType **) Result2,
-                               (LineType **) Result3))
-    return (LINE_TYPE);
+      && SearchLineByLocation (locked, (LayerType **)Result1, (LineType **) Result2, (LineType **) Result3))
+    return LINE_TYPE;
 
     if ((HigherAvail & (PIN_TYPE | PAD_TYPE)) == 0 &&
       Type & ARCPOINT_TYPE &&
-      SearchArcPointByLocation (locked,
-                                Result1,
-                                (ArcType **) Result2,
-                                (PointType **) Result3))
-    return (ARCPOINT_TYPE);
+      SearchArcPointByLocation (locked, (LayerType **)Result1, (ArcType **) Result2, (PointType **) Result3))
+    return ARCPOINT_TYPE;
 
   if ((HigherAvail & (PIN_TYPE | PAD_TYPE)) == 0 && Type & ARC_TYPE &&
-      SearchArcByLocation (locked,
-                           Result1,
-                           (ArcType **) Result2,
-                           (ArcType **) Result3))
-    return (ARC_TYPE);
+      SearchArcByLocation (locked, (LayerType **)Result1, (ArcType **) Result2, (ArcType **) Result3))
+    return ARC_TYPE;
 
   if ((HigherAvail & (PIN_TYPE | PAD_TYPE)) == 0 && Type & TEXT_TYPE
-      && SearchTextByLocation (locked,
-                               Result1,
-                               (TextType **) Result2,
-                               (TextType **) Result3))
-    return (TEXT_TYPE);
+      && SearchTextByLocation (locked, (LayerType **)Result1, (TextType **) Result2, (TextType **) Result3))
+    return TEXT_TYPE;
 
   if (Type & POLYGON_TYPE &&
-      SearchPolygonByLocation (locked,
-                               Result1,
-                               (PolygonType **) Result2,
-                               (PolygonType **) Result3))
+      SearchPolygonByLocation (locked, (LayerType **)Result1, (PolygonType **) Result2, (PolygonType **) Result3))
     {
       if (HigherAvail)
         {
@@ -1195,14 +1181,20 @@ SearchLayerObjectInternal (LayerType *SearchLayer,
             &(*(PolygonType **) Result2)->BoundingBox;
           double area =
             (double) (box->X2 - box->X1) * (double) (box->X2 - box->X1);
+          /* XXX: BEHAVIOURAL CHANGE */
           if (HigherBound < area)
-            break;
+//            break;
+            return NO_TYPE;
           else
-            return (POLYGON_TYPE);
+            return POLYGON_TYPE;
         }
       else
-        return (POLYGON_TYPE);
+        {
+          return POLYGON_TYPE;
+        }
     }
+
+  return NO_TYPE;
 }
 
 /* ---------------------------------------------------------------------------
@@ -1228,7 +1220,7 @@ SearchObjectByLocation (unsigned Type,
   void **pr1 = &r1, **pr2 = &r2, **pr3 = &r3;
   Cardinal i;
   double HigherBound = 0;
-  int HigherAvail = NO_TYPE;
+  unsigned int HigherAvail = NO_TYPE;
   int locked = Type & LOCKED_TYPE;
   unsigned int type;
   /* setup variables used by local functions */
@@ -1311,28 +1303,28 @@ SearchObjectByLocation (unsigned Type,
       HigherAvail = ELEMENT_TYPE;
     }
 
-  for (i = -2; i < max_copper_layer + 2; i++)
+  type = SearchLayerObjectInternal (&PCB->Data->SOLDERMASKLAYER, Type, HigherAvail, HigherBound, Result1, Result2, Result3);
+  if (type != NO_TYPE)
+    return type;
+
+  type = SearchLayerObjectInternal (&PCB->Data->SILKLAYER, Type, HigherAvail, HigherBound, Result1, Result2, Result3);
+  if (type != NO_TYPE)
+    return type;
+
+  for (i = 0; i < max_copper_layer; i++)
     {
-      if (i == -2)
-	SearchLayer = &PCB->Data->SOLDERMASKLAYER;
-      else if (i == -1)
-	SearchLayer = &PCB->Data->SILKLAYER;
-      else if (i < max_copper_layer)
-	SearchLayer = LAYER_ON_STACK (i);
-      else if (i == max_copper_layer)
-	{
-	  SearchLayer = &PCB->Data->BACKSILKLAYER;
-	  if (!PCB->InvisibleObjectsOn)
-	    continue;
-	}
-      else if (i == max_copper_layer + 1)
-	{
-	  SearchLayer = &PCB->Data->BACKSOLDERMASKLAYER;
-	  if (!PCB->InvisibleObjectsOn)
-	    continue;
-        }
+      type = SearchLayerObjectInternal (LAYER_ON_STACK (i), Type, HigherAvail, HigherBound, Result1, Result2, Result3);
+      if (type != NO_TYPE)
+        return type;
+    }
+
+  if (PCB->InvisibleObjectsOn)
+    {
+      type = SearchLayerObjectInternal (&PCB->Data->BACKSILKLAYER, Type, HigherAvail, HigherBound, Result1, Result2, Result3);
+      if (type != NO_TYPE)
+        return type;
 
-      type = SearchLayerObjectInternal (SearchLayer, Result1, Result2, Result3);
+      type = SearchLayerObjectInternal (&PCB->Data->BACKSOLDERMASKLAYER, Type, HigherAvail, HigherBound, Result1, Result2, Result3);
       if (type != NO_TYPE)
         return type;
     }
